File: backend/src/agents/agentRunner.ts
import { db } from '../shared/db/pool';
import { runSupervisorAgent } from './supervisorAgent';

/**
 * Recover agent runs that were left in RUNNING state when the process previously crashed.
 * Called once at startup. Marks each stuck run as DEAD so operators can inspect them
 * and new runs can be enqueued without confusion.
 *
 * A run is "stuck" when:
 *   status = 'RUNNING'  AND  started_at < now() - timeout_seconds interval
 *
 * 'DEAD' is used (not 'FAILED') to distinguish crash-recovery from normal failure,
 * matching the Day 3 durable execution status vocabulary.
 */
export async function recoverDeadRuns(): Promise<void> {
  const { rows } = await db.query<{ id: string }>(
    `UPDATE agent.agent_runs ar
     SET status             = 'DEAD',
         termination_reason = 'PROCESS_CRASH_RECOVERY',
         completed_at       = now()
     FROM agent.agent_definitions ad
     WHERE ar.agent_definition_id = ad.id
       AND ar.status = 'RUNNING'
       AND ar.started_at < now() - (ad.timeout_seconds * INTERVAL '1 second')
     RETURNING ar.id`
  );
  if (rows.length > 0) {
    console.warn(
      `[agentRunner] recoverDeadRuns: marked ${rows.length} stuck run(s) as DEAD`,
      rows.map(r => r.id)
    );
  }
}

export async function executeAgentRun(runId: string): Promise<void> {
  // Load the run joined with its supervisor agent definition
  const { rows: runRows } = await db.query(
    `SELECT ar.id, ar.student_id, ar.goal_snapshot, ar.triggered_by_user_id,
            ar.status, ar.agent_definition_id,
            ad.max_steps, ad.max_tool_calls, ad.timeout_seconds
     FROM agent.agent_runs ar
     JOIN agent.agent_definitions ad ON ad.id = ar.agent_definition_id
     WHERE ar.id = $1`,
    [runId]
  );

  if (runRows.length === 0) return;  // run not found — nothing to do
  const run = runRows[0];

  // Idempotency: only execute QUEUED runs
  if (run.status !== 'QUEUED') return;

  // Load the specialist agent definition
  const { rows: specRows } = await db.query(
    `SELECT id, max_steps, max_tool_calls, timeout_seconds
     FROM agent.agent_definitions
     WHERE name = 'learning_specialist_agent' AND is_active = true
     ORDER BY version DESC LIMIT 1`
  );

  if (specRows.length === 0) {
    await db.query(
      `UPDATE agent.agent_runs
       SET status='FAILED', termination_reason='SPECIALIST_DEF_NOT_FOUND', completed_at=now()
       WHERE id=$1`,
      [runId]
    );
    return;
  }

  // CAS update — prevents double-execution if two workers race
  const { rowCount } = await db.query(
    `UPDATE agent.agent_runs
     SET status='RUNNING', started_at=now()
     WHERE id=$1 AND status='QUEUED'`,
    [runId]
  );

  if ((rowCount ?? 0) === 0) return;  // another worker already took it

  try {
    await runSupervisorAgent({
      supervisorRunId:   runId,
      studentId:         run.student_id,
      goal:              run.goal_snapshot,
      triggeredByUserId: run.triggered_by_user_id,
      specDef:           specRows[0],
    });

    await db.query(
      `UPDATE agent.agent_runs
       SET status='SUCCEEDED', termination_reason='LearningPlanPersisted', completed_at=now()
       WHERE id=$1`,
      [runId]
    );
  } catch (err) {
    const reason = err instanceof Error ? err.message : 'Unknown error';
    await db.query(
      `UPDATE agent.agent_runs
       SET status='FAILED', termination_reason=$2, completed_at=now()
       WHERE id=$1`,
      [runId, reason.slice(0, 500)]
    );
  }
}

File: ai-service/app/agents/agent_runner.py
from __future__ import annotations

import warnings

from app.agents.supervisor_agent import run_supervisor_agent
from app.repositories import agent_repository


def execute_agent_run(run_id: str) -> None:
    """Load a QUEUED agent run, execute it, and mark it SUCCEEDED or FAILED."""
    run = agent_repository.load_run(run_id)
    if run is None:
        return
    if run.get("status") != "QUEUED":
        return

    spec_def = agent_repository.load_specialist_def()
    if spec_def is None:
        agent_repository.update_run_status(run_id, "FAILED", "SPECIALIST_DEF_NOT_FOUND")
        return

    acquired = agent_repository.cas_start_run(run_id)
    if not acquired:
        return

    try:
        run_supervisor_agent(
            supervisor_run_id=run_id,
            student_id=run["student_id"],
            goal=run["goal_snapshot"],
            triggered_by_user_id=run.get("triggered_by_user_id"),
            spec_def=spec_def,
        )
        agent_repository.update_run_status(run_id, "SUCCEEDED", "LearningPlanPersisted")
    except Exception as err:
        reason = str(err)[:500]
        agent_repository.update_run_status(run_id, "FAILED", reason)


def recover_dead_runs() -> list[str]:
    """Mark stuck RUNNING runs as DEAD on service startup.

    A run is stuck when its started_at exceeds the agent definition's timeout.
    Uses 'DEAD' (not 'FAILED') to distinguish crash recovery from normal failure.
    """
    recovered = agent_repository.recover_dead_runs()
    if recovered:
        warnings.warn(
            f"[agent_runner] recover_dead_runs: marked {len(recovered)} stuck run(s) as DEAD: {recovered}",
            RuntimeWarning,
            stacklevel=1,
        )
    return recovered
